lib/spack/spack/util/download.py

```diff
@@ -0,0 +1,345 @@
+        yield UrllibStream(url=url, response=response)
+
+
+class CurlStream(UrlStream):
+    """Streams a file from a URL using curl."""
+
+    def __init__(self, *, url: str, headers: HTTPMessage, effective_url: str, stream):
+        self.url = url
+        self.headers = headers
+        self._effective_url = effective_url
+        self._stream = stream
+
+    def read(self, size: int) -> bytes:
+        return self._stream.read(size)
+
+    def geturl(self) -> str:
+        return self._effective_url
+
+
+def curl_cookie_args(cookie) -> List[str]:
+    """Returns the arguments to pass to curl to use a cookie.
+
+    This can be used as a helper function to generate the arguments for the constructor.
+    """
+    return ["-j", "-b", cookie]
+
+
+@contextlib.contextmanager
+def curl_stream(
+    *, url: str, timeout: int, config_args: Optional[List[str]] = None
+) -> Generator[UrlStream, Any, None]:
+    config_args = config_args or []
+    cmd = [which_string("curl", required=True)]
+    cmd.extend(base_curl_fetch_args(url, timeout, status_bar=False, user_agent=SPACK_USER_AGENT))
+    cmd.extend(config_args)
+    with _start_curl_process(cmd) as curl_process:
+        scheme = urllib.parse.urlparse(url).scheme
+        assert curl_process.stdout is not None, "curl process stdout is None"
+        stream = curl_process.stdout
+        effective_url = url
+        if scheme not in ("http", "https"):
+            headers = HTTPMessage()
+        else:
+            # curl echoes intermediate redirect responses, so we might get multiple responses
+            finished = False
+            while not finished:
+                headers, effective_url, finished = _next_http_headers(
+                    url=effective_url, stream=stream
+                )
+        yield CurlStream(url=url, headers=headers, effective_url=effective_url, stream=stream)
+    check_curl_code(curl_process.returncode)
+
+
+@contextlib.contextmanager
+def _start_curl_process(curl_cmd):
+    with subprocess.Popen(
+        curl_cmd, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL
+    ) as curl_process:
+        yield curl_process
+
+
+def _next_http_headers(url, stream) -> Tuple[http.client.HTTPMessage, str, bool]:
+    """Returns the next headers from the stream."""
+
+    finished = True
+    status_line = stream.readline().decode("iso-8859-1")
+    if not status_line.startswith("HTTP/"):
+        raise FetchError(f"Failed to fetch {url}: unexpected status line: {status_line}")
+    try:
+        status = int(status_line.split()[1])
+    except ValueError:
+        raise FetchError(
+            f"Failed to fetch {url}: cannot parse HTTP status code from {status_line}"
+        )
+
+    headers = http.client.parse_headers(stream)
+    if 400 <= status < 600:
+        raise DetailedHTTPError(
+            urllib.request.Request(url), status, HTTPStatus(status).phrase, headers, None
+        )
+
+    elif 300 <= status < 400:
```

Comment:
The only HTTP status code that is in the 300-399 range which is final but not a redirect is `304 Not Modified`.

It's used for conditional requests, see `spack.binary_distribution.DefaultIndexFetcher.conditional_fetch` where we do `GET ...` with `If-None-Match: "<etag>"` and you expect either 2XX with contents or `3XX` with `304 Not Modified`.